kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,700 @@
+	defer uruntime.HandleCrash()
+
+	// Start the Kubernetes informer, which will start syncing with the Kubernetes API.
+	log.Info("Starting LoadBalancer controller")
+	go c.informer.Run(stopCh)
+
+	// Wait until we are in sync with the Kubernetes API before starting the
+	// resource cache.
+	log.Debug("Waiting to sync with Kubernetes API (Service)")
+	for !c.informer.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+	}
+	log.Debug("Finished syncing with Kubernetes API (Service)")
+	c.lbResourceCache.Run(timer.String())
+
+	for i := 0; i < c.cfg.NumberOfWorkers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	log.Info("LoadBalancer controller is now running")
+
+	// Register to the syncer datafeed for updates to ippools and ipamBlocks
+	c.RegisterWith(c.dataFeed)
+	c.dataFeed.Start()
+	go c.acceptScheduledRequests(stopCh)
+
+	<-stopCh
+	log.Info("Stopping Service controller")
+}
+
+func (c *loadBalancerController) RegisterWith(f *node.DataFeed) {
+	f.RegisterForNotification(model.BlockKey{}, c.onUpdate)
+	f.RegisterForNotification(model.ResourceKey{}, c.onUpdate)
+	f.RegisterForSyncStatus(c.onStatusUpdate)
+}
+
+func (c *loadBalancerController) onStatusUpdate(s bapi.SyncStatus) {
+	c.syncerUpdates <- s
+}
+
+func (c *loadBalancerController) onUpdate(update bapi.Update) {
+	switch update.KVPair.Key.(type) {
+	case model.ResourceKey:
+		switch update.KVPair.Key.(model.ResourceKey).Kind {
+		case api.KindIPPool:
+			c.syncerUpdates <- update.KVPair
+		}
+	case model.BlockKey:
+		c.syncerUpdates <- update.KVPair
+	default:
+		log.Warnf("Unexpected kind received over syncer: %s", update.KVPair.Key)
+	}
+}
+
+func (c *loadBalancerController) acceptScheduledRequests(stopCh <-chan struct{}) {
+	log.Infof("Will run periodic IPAM sync every %s", timer)
+	t := time.NewTicker(timer)
+	for {
+		select {
+		case update := <-c.syncerUpdates:
+			c.handleUpdate(update)
+		case <-t.C:
+			log.Infof("Running periodic IPAM sync of Service LoadBalancer")
+			c.syncIPAM()
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (c *loadBalancerController) handleUpdate(update interface{}) {
+	switch update := update.(type) {
+	case bapi.SyncStatus:
+		c.syncStatus = update
+		switch update {
+		case bapi.InSync:
+			log.WithField("status", update).Info("Syncer is InSync, kicking sync channel")
+			kick(c.syncChan)
+		}
+		return
+	case model.KVPair:
+		switch update.Key.(type) {
+		case model.ResourceKey:
+			switch update.Key.(model.ResourceKey).Kind {
+			case api.KindIPPool:
+				log.Infof("Running periodic IPAM sync of Service LoadBalancer")
+				c.handleIPPoolUpdate(update)
+				return
+			}
+		case model.BlockKey:
+			c.handleBlockUpdate(update)
+			return
+		}
+	}
+
+}
+
+func (c *loadBalancerController) handleBlockUpdate(kvp model.KVPair) {
+	if kvp.Value != nil {
+		host := kvp.Value.(*model.AllocationBlock).Affinity
+		if host != nil && *host == fmt.Sprintf("host:%s", api.VirtualLoadBalancer) {
```

Comment:
For virtual affinities, we really want to be using something other than "host"

i.e., the IPAM code should be writing affinities like this for LB IPs:

```
affinity: "virtual:loadbalancer"
```

And this for "real" IPs:

```
affinity: "host:node"
```

The "prefix" before the `:` is meant to indicate what type of thing this is affine to. If we just use `host:` like we do for real IPs, then anyone can create a real Node in the cluster named `virtual-load-balancer` and it will confuse our code.